incubator/hnc/pkg/validators/hierarchy.go

```diff
@@ -151,13 +166,40 @@ func (v *Hierarchy) checkForest(hc *api.HierarchyConfiguration) ([]authzReq, adm
 	// have visibility into its ancestry and descendents, and this check can only fail if the new
 	// parent conflicts with something in the _existing_ hierarchy.
 	if reason := ns.CanSetParent(newParent); reason != "" {
-		return nil, deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
+		return deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
 	}
 
-	// TODO: requiredChild check: https://github.com/kubernetes-sigs/multi-tenancy/issues/99
+	// Prevent changing parent of a required child
+	if ns.RequiredChildOf != "" && ns.RequiredChildOf != newParent.Name() {
+		reason := fmt.Sprintf("Cannot set the parent of %q to %q because it's required child of %q", ns.Name(), newParent.Name(), ns.RequiredChildOf)
+		return deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
+	}
 
-	// The structure looks good. Get the list of namespaces we need authz checks on.
-	return v.needAuthzOn(curParent, newParent), allow("")
+	return allow("")
+}
+
+func (v *Hierarchy) checkRequiredChildren(ns *forest.Namespace, requiredChildren []string) admission.Response {
+	for _, child := range requiredChildren {
+		cns := v.Forest.Get(child)
+		// Early exit if we find the newly-added requiredChild
+		if !cns.Exists() {
+			return allow("RequiredChild namesapce is just created")
+		}
+		if cns.RequiredChildOf != "" && cns.RequiredChildOf != ns.Name() {
+			reason := fmt.Sprintf("Cannot set %q as the required child of %q because it's a required child of %q", cns.Name(), ns.Name(), cns.RequiredChildOf)
+			return deny(metav1.StatusReasonConflict, "Illegal requiredChild: "+reason)
+		}
+		if cns.Parent() != nil && cns.Parent() != ns {
+			reason := fmt.Sprintf("Cannot set %q as the required child of %q because it's already a child of %q", cns.Name(), ns.Name(), cns.Parent().Name())
+			return deny(metav1.StatusReasonConflict, "Illegal requiredChild: "+reason)
+		}
+		if cns.Parent() == nil {
+			reason := fmt.Sprintf("Cannot set %q as the required child of %q because it's a root", cns.Name(), ns.Name())
+			return deny(metav1.StatusReasonConflict, "Illegal requiredChild: "+reason)
+		}
```

Comment:
Should we replace all of these with "because it already exists and is not a child of X"?